File: protocol/src/main/java/com/sudox/protocol/client/network/SocketClient.java
package com.sudox.protocol.client.network;

import androidx.annotation.NonNull;

import java.nio.ByteBuffer;

public class SocketClient {

    private final long pointer;

    public SocketClient(@NonNull String host, short port) {
        System.loadLibrary("ssockets");
        this.pointer = createNativeInstance(host, port);
    }

    public interface ClientCallback {
        void socketConnected();

        void socketClosed(boolean error);

        void socketReceive();
    }

    public void connect() {
        connect(pointer);
    }

    public void close(boolean error) {
        if (opened()) {
            close(pointer, error);
        }
    }

    public boolean opened() {
        return opened(pointer);
    }

    public void sendBuffer(@NonNull ByteBuffer buffer) {
        if (buffer.isDirect()) {
            sendBuffer(pointer, buffer, buffer.limit());
        } else {
            send(pointer, buffer.array());
        }
    }

    public void send(@NonNull byte[] data) {
        send(pointer, data);
    }

    @NonNull
    public byte[] readBytes(int count) {
        return readBytes(pointer, count);
    }

    /**
     * Returns count of read bytes.
     * Returns -1 if error thrown.
     * <p>
     * P.S.: Supports only direct buffers!
     */
    public int readToByteBuffer(@NonNull ByteBuffer buffer, int count, int offset) {
        int read = readToBuffer(pointer, buffer, count, offset);

        if (read > 0) {
            buffer.position(buffer.position() + read);
        }

        return read;
    }

    /**
     * Returns count of available bytes.
     * Returns -1 if error thrown.
     */
    public int availableBytes() {
        return availableBytes(pointer);
    }

    public void callback(@NonNull ClientCallback callback) {
        callback(pointer, callback);
    }

    private native long createNativeInstance(String host, short port);

    private native void connect(long pointer);

    private native void close(long pointer, boolean error);

    private native boolean opened(long pointer);

    private native void send(long pointer, byte[] data);

    private native void sendBuffer(long pointer, ByteBuffer buffer, int length);

    private native byte[] readBytes(long pointer, int count);

    private native int readToBuffer(long pointer, ByteBuffer buffer, int count, int offset);

    private native int availableBytes(long pointer);

    private native void callback(long pointer, ClientCallback callback);
}

File: protocol/src/main/java/com/sudox/protocol/client/network/SocketClient.kt
package com.sudox.protocol.client.network

import java.nio.ByteBuffer

class SocketClient(host: String, port: Short) {

    private var pointer: Long = 0

    init {
        loadLibrary()
        pointer = createNativeInstance(host, port)
    }

    fun connect() = connect(pointer)
    fun close(error: Boolean) = close(pointer, error)
    fun callback(callback: SocketCallback) = callback(pointer, callback)
    fun available() = available(pointer)
    fun send(buffer: ByteBuffer, urgent: Boolean) = send(pointer, buffer, buffer.limit(), urgent)
    fun read(count: Int): ByteArray = read0(pointer, count)

    fun read(buffer: ByteBuffer, count: Int, offset: Int): Int {
        val read = read1(pointer, buffer, count, offset)

        if (read > 0) {
            buffer.position(buffer.position() + read)
        }

        return read
    }
}

internal var isLibraryLoaded: Boolean = false

internal fun loadLibrary() {
    if (!isLibraryLoaded) {
        System.loadLibrary("ssockets")
    }
}

internal external fun createNativeInstance(host: String, port: Short): Long
internal external fun connect(pointer: Long)
internal external fun close(pointer: Long, error: Boolean)
internal external fun callback(pointer: Long, callback: SocketCallback)
internal external fun available(pointer: Long): Int
internal external fun send(pointer: Long, buffer: ByteBuffer, length: Int, urgent: Boolean)
internal external fun read0(pointer: Long, count: Int): ByteArray
internal external fun read1(pointer: Long, buffer: ByteBuffer, count: Int, offset: Int): Int